tuner/bayes_opt_gp/dffml_tuner_bayes_opt_gp/bayes_opt_gp.py

```diff
@@ -0,0 +1,166 @@
+        train_data: Union[BaseSource, Record, Dict[str, Any]],
+        test_data: Union[BaseSource, Record, Dict[str, Any]],
+    ):
+        """
+        Method to optimize hyperparameters by Bayesian optimization using Gaussian Processes
+        as the surrogate model.
+        Uses a grid of hyperparameters in the form of a dictionary present in config,
+        Trains each permutation of the grid of parameters and compares accuracy.
+        Sets model to the best parameters and returns highest accuracy.
+
+        Note that for this tuner, each hyperparameter field to be tuned must have exactly 2 values 
+        specified, representing the minimum and maximum values in the search space for that 
+        hyperparameter. Additionally, they must be either float/integer values. Otherwise,
+        an error is raised.
+
+        Parameters
+        ----------
+        model : ModelContext
+            The Model which needs to be used.
+
+        feature : Feature
+            The Target feature in the data.
+
+        accuracy_scorer: AccuracyContext
+            The accuracy scorer that needs to be used.
+
+        train_data: SourcesContext
+            The train_data to train models on with the hyperparameters provided.
+
+        sources : SourcesContext
+            The test_data to score against and optimize hyperparameters.
+
+        Returns
+        -------
+        float
+            The highest score value
+        """
+
+        nest_asyncio.apply()
+
+        def check_parameters(pars):
+            for (pax, vals) in pars.items():
+                if len(vals) != 2:
+                    raise InvalidParametersException(
+                        f"2 values are not provided for parameter {pax}"
+                    )
+                for val in vals:
+                    if not type(val) is float and not type(val) is int:
+                        raise InvalidParametersException(
+                            f"Parameter {pax} is not of type int or float."
+                        )
+            return True
+
+        check_parameters(self.parent.config.parameters)
+
+        logging.info(
+            f"Optimizing model with Bayesian optimization with gaussian processes: {self.parent.config.parameters}"
+        )
+
+        def func(**vals):
+            with model.parent.config.no_enforce_immutable():
+                for param in vals.keys():
+
+                    if (
+                        hasattr(model.parent.config, param)
+                        and model.parent.config.__annotations__[param].__name__
+                        == "int"
+                    ):
+                        setattr(model.parent.config, param, int(vals[param]))
+                    else:
+                        setattr(model.parent.config, param, vals[param])
+
+                train(model.parent, *train_data)
+                acc = score(model.parent, accuracy_scorer, feature, *test_data)
```

Comment:
```suggestion
                await train(model.parent, *train_data)
                acc = await score(model.parent, accuracy_scorer, feature, *test_data)
```

Reply:
Unfortunately, I don't think it's possible to make these functions asynchronous. The Bayesian optimizer requires that the score of the previous iteration be available to update its prior and start the next iteration, which asynchronicity cannot guarantee. Specifically, when I try to make them async, I run into this error: 
```
TypeError: '>' not supported between instances of 'coroutine' and 'coroutine'
```
which makes sense since each iteration is now an asynchronous co-routine that  may not necessarily have completed at the time of comparison. So for this instance, I think it is necessary that they are synchronous.